src/GafferImage/OpenImageIOReader.cpp

```diff
@@ -493,206 +920,371 @@ class File
+					tileChannelPointers[ subIndex ] = &tileAlloc->writable()[0];
+					resultChannels->members()[ subIndex ] = std::move( tileAlloc );
+				}
+			}
+			else
+			{
+				// For deep images, we allocate the sample offsets, but can't allocate any of the channel
+				// data until we know how many samples there are
+				resultOffsets = new ObjectVector();
+				resultOffsets->members().resize( tileBatchNumTiles );
+				tileOffsetPointers.resize( tileBatchNumTiles );
 
-			// Pull data apart into tiles ( separate for each channel instead of interleaved )
-			int tileBatchNumElements = nchannels * view.tileBatchSize.y * view.tileBatchSize.x;
-			ObjectVectorPtr result = new ObjectVector();
-			result->members().resize( tileBatchNumElements );
+				for( int subIndex = 0; subIndex < tileBatchNumTiles; subIndex++ )
+				{
+					IntVectorDataPtr tileAlloc = new IECore::IntVectorData();
+					podVectorResizeUninitialized<int>( tileAlloc->writable(), ImagePlug::tilePixels() );
 
-			ObjectVectorPtr resultChannels = result;
+					tileOffsetPointers[ subIndex ] = &tileAlloc->writable()[0];
+					resultOffsets->members()[ subIndex ] = std::move( tileAlloc );
+				}
+			}
 
-			std::vector< int > deepTileSizes;
+			// Convert the data window from the file into Gaffer coordinates
+			const V2i fileDataOrigin( spec.x, spec.y );
+			const Box2i gafferDataWindow = flopDisplayWindow(
+				Box2i( fileDataOrigin, fileDataOrigin + V2i( spec.width, spec.height ) ),
+				spec.full_y, spec.full_height
+			);
+
+			// The region of each tile that is within the data window
+			const std::vector< Box2i > tileDataWindows = calculateTileDataWindows(
+				tileBatchNumTiles, tileBatchOrigin, view.tileBatchSize, gafferDataWindow
+			);
+
+			// Find the portion of the data window that intersects with the current tile batch,
+			// and convert it from Gaffer coordinates to file coordinates.
+			const V2i tileBatchOriginXY( tileBatchOrigin.x, tileBatchOrigin.y );
+			const Box2i targetRegion = BufferAlgo::intersection(
+				Box2i( tileBatchOriginXY, tileBatchOriginXY + view.tileBatchSize * ImagePlug::tileSize() ),
+				gafferDataWindow
+			);
+			const Box2i fileTargetRegion = flopDisplayWindow(
+				targetRegion, view.imageSpec.full_y, view.imageSpec.full_height
+			);
+
+			// It would probably be more efficient if we just did two separate traversals of the input regions,
+			// with the first one setting EXR_DECODE_SAMPLE_DATA_ONLY, rather than decoding everything up front,
+			// and having to hold it in memory while we compute our sample offsets before we can actually use it.
+			// OIIO does not expose any way to do this though ... it would really make sense if passing chend = 0
+			// would get you just the sample counts.
+			std::vector< OIIO::DeepData > deepRectsData;
+			std::vector< Box2i > deepRects;
+
+			bool usingExrCore = strcmp( m_imageInput->format_name(), "openexr" ) == 0 && !g_disableExrCore;
```

Comment:
> Are there any other formats where we might benefit from the threaded reads?

The main one that seems like it would be useful would be Tiff, if OIIO itself didn't hold the lock while decompressing, as you observed.